### ISSUE TITLE
Improve Config System with Registry Pattern

### ISSUE THREAD

**Issue by Skylion007** [rosinality/tensorfn#1]
Is this repo not update to date with pip?

The version number of this repo is behind the pip version and I have tried running the code from https://github.com/rosinality/denoising-diffusion-pytorch/ but that only seems to work with the pip copy of tensorfn not the one built form here. I get key errors for the phase 'flat' not being defined. Also the pip version is 0.1.4 vs 0.1 

```
 in make
    self.decay,
  File "/private/home/$USER/git/denoising-diffusion-pytorch/tensorfn_git/tensorfn/optim/lr_scheduler.py", line 158, in cycle_scheduler
    return PhaseScheduler(optimizer, phases)
  File "/private/home/$USER/git/denoising-diffusion-pytorch/tensorfn_git/tensorfn/optim/lr_scheduler.py", line 73, in __init__
    phase_fn = phase_map[phase_name]
KeyError: 'flat'
```
happens with the default diffussion.conf in that repo

Also, it is a bit annoying we can't register optimizers without directly editing this library. It would be nice if you added a registry syntax that automatically could add new optimizers to the list of supported ones.

**Comment by rosinality** [comment]
I have updated the repo. Sorry for inconvenience.

You can add optimizers without modify the library itself. It would be like this:

```python
from typing import Union

from pydantic import validator, StrictStr, StrictFloat
from tensorfn.config import Config, Optimizer

class NewOptimizer(Config):
    type: StrictStr

    lr: StrictFloat = 0.001
    betas: Tuple[StrictFloat, StrictFloat] = (0.9, 0.999)
    eps: StrictFloat = 1e-6
    weight_decay: StrictFloat = 0

    @validator("type")
    def check_type(cls, v):
        if v != "new_optimizer":
            raise ValueError("Optimizer type not match for new optimizer")

        return v

    def make(self, params):
        return tensor_optim.LAMB(
            params, self.lr, self.betas, self.eps, self.weight_decay
        )

CustomOptimizer = Union[Optimizer, NewOptimizer]
```

**Comment by Skylion007** [comment]
Ah, I thought it was enumerating the list of optimizers based on the type annotation, but it does so by figuring out what extends config???

**Comment by rosinality** [comment]
Yes, it is enumerating the list of optimizers that could be accepted for the config. So if you extend the type using Union then you can make new optimizers to be available for the config.

**Comment by Skylion007** [comment]
I see, but where would the CustomOptimizer type actually be used?

**Comment by rosinality** [comment]
You can replace Optimizer in this line https://github.com/rosinality/denoising-diffusion-pytorch/blob/master/config.py#L25 to CustomOptimizer. Then config will accept new optimizer.

**Comment by Skylion007** [comment]
Yeah but using a registry-decorator pattern would mean that there would be no additional code changes needed on the user side. Simply importing the optimizer would add it to the list of available options. That's what we use Habitat-Lab: https://github.com/facebookresearch/habitat-lab See examples/new_actions.py

**Comment by rosinality** [comment]
Yes, this is experimental attempt to make config system that do type check & validation. I will try to improve this with registry pattern.
